### ISSUE TITLE
SSH Agent Forwarding

### ISSUE THREAD

**Issue by dt** [mobile-shell/mosh#120]
Would be great to be able to forward local SSH agents. Seems to ignore `ForwardAgent yes` in `~/.ssh/config`

**Comment by termie** [comment]
+1

**Comment by xtimmu** [comment]
+1

**Comment by NoUseFreak** [comment]
:thumbsup:

**Comment by benlemasurier** [comment]
+1

**Comment by timesking** [comment]
+1

**Comment by jravetch** [comment]
:+1:

**Comment by takaidohigasi** [comment]
+1

**Comment by RubenKelevra** [comment]
+1

**Comment by codeshrew** [comment]
:+1:

**Comment by aaroncline** [comment]
:+1:

**Comment by PAStheLoD** [comment]
Hey,

Since the code works (thanks to kevinr), I couldn't stand it just sitting somewhere in a lonely Git repo. So, here's a Debian (actually Ubuntu) package, that works (for me at least): https://github.com/PAStheLoD/mosh/tree/master#but-if-you-do

**Comment by joshSzep** [comment]
+1

**Comment by dejadejoder** [comment]
I may have founded a workaround using 'tmux':
1. login with normal ssh client
2. open a tmux terminal and detach it
3. login with mosh
4. attach previously created tmux terminal
You may be able to log with ssh-agent to your remote machines. I am not sure how long tmux will keep link with the ssh-agent...

**Comment by patcon** [comment]
Just as one sample use-case that would seem to not work with the `ProxyCommand` solution:

When working in a country with poor internet connection, and wanting to treat your local workstation as a stupid machine that SSH's into your actual workstation on a cloud provider. Using agent forwarding seems to make it as if my SSH key were deleted each time I log out

**Comment by ddarbyson** [comment]
+1

**Comment by davydovmax** [comment]
Must have feature for me (can't use mosh without it).

**Comment by rinne** [comment]
The current version of agent patches on top of the current upstream master branch of mosh is in #696. Some maintainer asked for patches to be re-merged and I assumed that it meant that they were going to pull the feature to mainline. However it has been 17 days (after I re-merged the feature) and nothing has happened since.
For now, cloning mosh sources from https://github.com/rinne/mosh and compiling agent-forwarding-merge-20151128 branch (--with-agent-forwarding), does the trick.

**Comment by Omeryl** [comment]
:+1:

**Comment by gavit** [comment]
Has it been added to the main yet?

**Comment by andr-04** [comment]
@rinne, I get the following error by using your brach:
```
bash-3.2$ mosh --forward-agent xxx
mosh-server: illegal option -- A
Usage: mosh-server new [-s] [-v] [-i LOCALADDR] [-p PORT[:PORT2]] [-c COLORS] [-l NAME=VALUE] [-- COMMAND...]
```

**Comment by rinne** [comment]
Three possible explanations.

1) You don't compile mosh from agent branch (agent-forwarding-merge-20151128), see #696 which is the current pull request.

2) You have different mosh installed in your system and you are actually running it.

3) You don't have agent enabled version installed in the server side in the system you are logging into. The error message makes this one the likely culprit.

**Comment by andr-04** [comment]
Do you mean that mosh-server from branch must be installed on the ssh-server as well?

**Comment by rinne** [comment]
Yes. Mosh server must of course support agent forwarding too. Just like in ssh connections server creates the proxy authentication socket in remote host and forwards the agent requests to through the ssh protocol connection back to the client and client in its turn forwards them to the local agent socket, which can of course be another proxy socket, if there are multiple chained connections.

**Comment by iurii-polishchuk** [comment]
Looking forward for merge into master.

**Comment by arlofaria** [comment]
+1

**Comment by fleshenough** [comment]
Come on, people.
Wake the fuck up.

Just chop this....

'-S', 'none'

out of the mosh script.

Works great.

I'm so tired of thinking for you all.

**Comment by dlangille** [comment]
Any news on this getting into mosh?  I cannot use it without ssh-agent.

**Comment by rinne** [comment]
@dlangille You might like to check out #696. The pull request there is up to date with current mosh master and while I don't hold my breath waiting it to be merged, you can compile from my forked code.

**Comment by skylerwshaw** [comment]
:+1:

**Comment by wagnerone** [comment]
Got excited about and really wanted to use Mosh. Spent a lot of time setting it up only to find agent forwarding doesn't work. Doh! Came here to find this discussion. Looking forward to a resolution!

**Comment by rinne** [comment]
@wagnerone Please check out #696.

**Comment by dlangille** [comment]
I appreciate the references to patches, but we need to get this into the code base before most people would consider using it.

**Comment by rinne** [comment]
Sure. Don't we all? The pointer was to more recent discussion and your upvotes and appreciations are appreciated there too. Lack of interest from the mainline developers has been quite ubiquitous all the way. I have been merging the mainline changes and resolving the conflicts every now and then in hope that they might merge the patches some day.

**Comment by dlangille** [comment]
Lack of developer interest is disappointing. Is it an objection to the feature, interests lie elsewhere, or complete silence on the topic?

**Comment by rinne** [comment]
My interpretation is that there is some small objection to the feature for the reason unclear (mainly "I don't use it, so it must not be necessary"). Secondly the patch is quite big, because mosh as it was (and is) wasn't really made to accommodate such extensions. About the actual implementation I've seen only minor and largely irrelevant discontent. I remember that @keithw wasn't originally sure, whether the way out-of-bans protocol (used by agent forwarding) data piggy-packs to mosh protocol packets is the right way to go, but that is so very basic decision in agent forwarding implementation that there isn't really too much to be done for that given that entire feature isn't going to be rewritten, which is something I absolutely won't do.

**Comment by yookoala** [comment]
Really like to see #696 merged.

Although I don't feel safe to store any SSH key on server, I need to occasionally do git pull PHP code base on server for deployments and upgrades. The lack of agent forwarding is a pain in ass.

**Comment by ei-grad** [comment]
I just use an extra pure-ssh connection to forward agent, when I need it, and do export SSH_AUTH_SOCK variable from that connection. And I'm happy.

**Comment by inertia186** [comment]
+1 to merge #696

**Comment by cgull** [comment]
I'd like to consider this for Mosh 1.3, after we release 1.2.6.  I've started a discussion on the mosh-devel mailing list; if you're interested please comment there.

**Comment by tnguyen14** [comment]
@cgull do you have the link to the mailing list thread?

**Comment by cgull** [comment]
http://mailman.mit.edu/pipermail/mosh-devel/2016-May/001298.html
You can also subscribe to the list from there.
The list is also archived by http://gmane.org and https://mail-archive.com.

**Comment by rinne** [comment]
@tnguyen14 http://mailman.mit.edu/mailman/listinfo/mosh-devel

I regret that I don't have time for that mailing list, but just a quick response to specific concern of @cgull about nasty agents/agent clients potentially blocking the mosh session.

It may be possible for that to happen, but thenagain if you run a nasty program in your remote shell, it can freeze your terminal connection using much simpler approach like sending a kill signal to your mosh-server process or sending random crap to your pty. From what I remember about writing that agent forwarding code a few years back, I at least tried to write it so that if agent connection data starts to get buffered or traffic that is tunneled doesn't look like legit agent protocol packets of reasonable length, the agent connection gets very easily disconnected in order to save the terminal connection from blocking. I think agent connections are also dropped, if mosh client goes to off-line state, because it's most likely a right thing to do in aboutn 100% of the cases.

That much said, I don't think the oob-protocol (the one I implemented and on top of which the agent connections are forwarded) is in its current state suitable for general purpose tunneling. I think much more fundamental changes to mosh would be needed for that and I don't think that's worth the effort. In order to do that so that it really works, there would be need for (sub-)data channels with channel specific flow control and buffer management. That stuff is actually a big part of ssh protocol implementation and to get that stuff really to work took number of years. Easily disconnecting forwarded agent sessions is just a substitute for flow control and while it's usable for this specific stuff, it's not usable for generic tunneling.

Generic tunneling would also introduce a whole set of new semantic problems. E.g. what to do to a tunneled tcp connection, if the mosh connection gets into "disconnected state". Normal tcp connection would use retransmits and eventually break up the connection, if no ack packets are received, but since the tunnel proxy would remain open indefinitely, this would not happen and the only thing taking place would be a blocking tcp socket.

Anyways, I'm glad to see that there are at least some signs of maybe mainlining the agent forwarding.

**Comment by mikalv** [comment]
+1

**Comment by mefyl** [comment]
Well, +1

**Comment by tnguyen14** [comment]
@cgull now that 1.2.6 is released, is there any estimate on how long 1.3 will take to get out of the door?

**Comment by ksanislo** [comment]
For anyone who wants to be able to use mosh with ssh-agent forwarding, you can work around this problem by adding the following to your ~/.profile or other location that will run on ssh connect.

`
find /tmp -maxdepth 2 -type s -name "agent*" -user ${USER} -printf "%T@ %p\n" 2>/dev/null \
 | sort -n \
 | tail -1 \
 | cut -d' ' -f2 \
 | xargs -rn1 -I '{}' ln -sf '{}' /home/${USER}/.ssh/ssh_auth_sock

export SSH_AUTH_SOCK=/home/${USER}/.ssh/ssh_auth_sock
`

I also put the following into my user's crontab, so the symlink will follow when the local socket changes, though I don't think this is actually required since the .profile will catch it on the new login:
`
*   *   *   *   *   find /tmp -maxdepth 2 -type s -name "agent*" -user \`whoami\` -printf "\%T@ \%p\n" 2>/dev/null | sort -n | tail -1 | cut -d' ' -f2 | xargs -rn1 -I '{}' ln -sf '{}' /home/\`whoami\`/.ssh/ssh_auth_sock
`
It's not perfect, since it requires extra things on the remote server, but it gets the job done pretty well.

This is the same hack used to get ssh-agent forwarding to work with tmux, and I found this ticket by accident and personally didn't even realize that mosh wasn't forwarding the agent itself since I was already using this setup.

**Comment by jorgelr** [comment]
@ksanislo With that setup, how do you connect? 

mosh bastion mosh internal_server?

**Comment by ksanislo** [comment]
@jorgelr I use mosh to my bastions, and typically just ssh to internal servers from there. Depending on what linux distro your bastion is running, you might need to tweak how you're making the symlinked ssh_auth_sock. 

The general idea is that you want the environment variable inside your connected mosh session to end up set like this:

$ echo $SSH_AUTH_SOCK
/home/kens/.ssh/ssh_auth_sock

$ readlink -f $SSH_AUTH_SOCK                                                                                                                                                                                                       
/tmp/ssh-Sq7wt8tgPj/agent.1151

So that symlink is pointed to the latest agent socket owned by your user, which should be the ssh connection that mosh created to connect. You'll still need ForwardAgent yes set in your .ssh/config file for the mosh bastion host.

**Comment by andersk** [comment]
I don’t want people to be misled here: @ksanislo’s suggestion requires that you leave open a separate SSH connection to the host in addition to the Mosh connection, to allow SSH to do the agent forwarding. The SSH connection that Mosh creates is used only for launching the Mosh server and is immediately closed.

**Comment by ksanislo** [comment]
Oh, you're right, the mosh session does go away instantly... I have a long running ssh between my home workstation and the bastion which is effectively where my ssh-agent ends up running... 

I suppose it's easy enough to duplicate until there's a proper solution though, just:

nohup ssh ${bastion_host} sleep 365d &

from somewhere that will stick around is enough to keep the agent available. You could get more creative and put it in a while true loop in case it gets disconnected.

**Comment by jorgelr** [comment]
Thanks -- I'll see if I can't get it working. But would absolutely send beer once this functionality get solved the right way in mosh. Crossing fingers and arms:)

**Comment by geeohgeegeeoh** [comment]
BSD find doesn't honour -printf. The recipe works if you modify it on BSD to use ls, and awk out the fields to get date sort order to work.

for the transient ssh thing, surely we want tail -2 | head -1 to get the *second last* ssh agent, which will be the long-lived instance?

find /tmp -maxdepth 2 -type s -name "agent*" -ls 2> /dev/null
 | awk '{print $8, $9, $10, $11}' \
 | sort -M \
 | tail -1 \

**Comment by glensc** [comment]
@geeohgeegeeoh `-1` for `tail` is deprecated with `coreutils`, does bsd `tail` support `-n1`? and perhaps it's just easier to use `gfind` instead than support ancient tools?

**Comment by geeohgeegeeoh** [comment]
the example in the discussion uses tail. so, I used what I saw. if you want
to deprecate tail, pass on up the foodchain because the original script for
.profile I modified used it.

good luck persuading BSD people to adopt gnuisms

**Comment by andersk** [comment]
`tail -n 1` is not a GNUism, it is [required by POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/tail.html) (where `tail -1` is not) and supported by  at least [GNU](https://www.gnu.org/software/coreutils/manual/coreutils.html#tail-invocation), [FreeBSD](http://mdoc.su/FreeBSD/tail.1), [NetBSD](http://mdoc.su/NetBSD/tail.1), [OpenBSD](http://mdoc.su/OpenBSD/tail.1), [DragonFly](http://mdoc.su/DragonFly/tail.1), and [Darwin](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/tail.1.html).

**Comment by maximvl** [comment]
So no updates, ticket is still open, what's the official status? Any solution besides hacks in comments above?

**Comment by HaleTom** [comment]
+1

**Comment by rinne** [comment]
@MaximF Well there is a solution https://github.com/mobile-shell/mosh/pull/696 :)

**Comment by MaximF** [comment]
Thank you for an update @rinne! Out of curiosity - is there an ETA for merging that PR?

**Comment by dlangille** [comment]
Point of order: #696 has existed since [at least] Nov 28, 2015, i.e about 14 months ago. There are several discussion points already in this thread. Search for 696 and read.

**Comment by geeohgeegeeoh** [comment]
@andersk this is all probably OBE. the gnuism comment was about gfind. I didn't realize the specificity of the -n1 thing. My solution only sometimes works, its bad code.  I'd far rather a clean solution which was generally applicable and portable was done, and I don't think coding to gnu find is that. POSIX behaviour is going to be fine, if tail -n1 is the syntax which is needed

**Comment by rlpowell** [comment]
On the point of security: my security team requires us to use a yubikey fob, which generates something like a one-time ssh key that is then required for login.  We then ssh to a bastion host, and outward from there to the real hosts.

This means that without ssh forwarding, I literally can't log in to anywhere except the bastion hosts, *for security reasons*.  This is a hard requirement from our security team.  So arguing that agent forwarding is bad because security cuts no mustard with me.

So, yeah, no mosh for me, I guess.

**Comment by dlangille** [comment]
Following on from the previous post: tools, not policy.

Let us decide if ssh-forwarding is for us or not.

**Comment by zenhack** [comment]
A few folks have posed similar (though, I think, not identical) workarounds, but here's what I've come up with: from one shell on the client:

    ssh -A ${host}

then, from another, log in with mosh:

    mosh ${host}

and in the mosh shell:

    export SSH_AUTH_SOCK=/tmp/ssh-<tab><tab>

...which should autocomplete to the socket forwarded by the first command. This doesn't allow roaming, since if the ssh connection breaks, you lose the agent, but it's good enough for getting a usable shell from a coffee shop's lousy wifi, which is my main use case. Would be nice to actually have the feature upstream though.

Ideally the first command could just be:

    ssh -AfqN ${host}

Which would obviate the need to keep an extra terminal open. But for some reason I haven't figured out, the auth socket doesn't show up when I do that.

Also: I am so glad github added the reactions feature -- cuts down on +1 noise.

**Comment by Zenexer** [comment]
I got about halfway through the comments before giving up.  Would someone mind summarizing where this issue stands?

Regarding the potential security issue: ssh-agent isn't just a program; it's also a protocol.  There are other programs that implement the same protocol with better security.  I use a YubiKey with the touch flag enabled; this means that each time a remote host wants to use my key, my YubiKey flashes, and I have about 15 seconds to touch it to authorize whatever's taking place.

The argument against this model, of course, is that there's no way for me to audit what's actually being authorized.  However, the same is perfectly true without ssh-agent forwarding.  You never have any guarantee that a given machine is doing what you tell it to do.  It's up to the end user to decide which risks are most significant in their environment in a case such as this, where there's no ubiquitous best practice.

**Comment by dlangille** [comment]
The previous post, in short, says: tools, not policy.

Let the users handle the policy.

**Comment by rinne** [comment]
@Zenexer It stands in #696. Basically a working patch set has existed for years but not merged to mainline. Meanwhile I've been keeping my agent enabled branch in sync with mainline master branch every now and then.

**Comment by bukzor** [comment]
I set up a bounty, because money talks, and put in the first $50 to get the ball rolling, I hope.

https://www.bountysource.com/issues/1587475-ssh-agent-forwarding

**Comment by MaximF** [comment]
@bukzor That is an amazing idea, thank you! Made my contribution there as well! :)

**Comment by rinne** [comment]
@bukzor @MaximF All find and well, but as stated, the actual code has already been written (by me about 5 years ago). There has been some mild critique over where and how it should be implemented, but there has been absolutely zero activity in reimplementing it in any other way it has already been done (and not merged to mainline). The current state of working patch is in #696.

**Comment by MaximF** [comment]
If it is already ready, why not to just merge it?
Better to have some functionality than no functionality :)

**Comment by rinne** [comment]
It is the full agent functionality. But merging is something that is out of my hands.

**Comment by MaximF** [comment]
Do you mean you don't have privileges to merge into a master branch? If so,
let's just make a fork! ;)

2017-08-29 23:25 GMT-07:00 Timo J. Rinne <notifications@github.com>:

**Comment by rinne** [comment]
There already is a fork that loosely follows the master branch of upstream mosh + agent patch. I won't commit myself to maintain it for the rest of my life though.

https://github.com/rinne/mosh

**Comment by bukzor** [comment]
@rinne Yes, this is about motivating upstream to take this seriously. There's actually two bounties now. A smaller one for this pull request in particular, and a larger for the original feature request.

[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=1587475)](https://www.bountysource.com/issues/1587475-ssh-agent-forwarding?utm_source=1587475&utm_medium=shield&utm_campaign=ISSUE_BADGE)

[![Bountysource](https://api.bountysource.com/badge/issue?issue_id=1587475)](https://www.bountysource.com/issues/1587475-ssh-agent-forwarding?utm_source=1587475&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by Apreche** [comment]
Just discovered Mosh today. Was excited to try it out. Discovered that Agent Forwarding didn't work. Connected to my server, and everything was great. Then I tried to git clone and realized there was no agent forwarding. Without that feature Mosh is completely useless to me. Saw that this issue is from 2012. That is very disappointing. Guess it's back to trusty old SSH for me.

**Comment by HaleTom** [comment]
If we have a working patch, what exactly is preventing the merge?

1. Enough time
2. Code quality
3. Design decisions
4. Other

What is blocking this so we can be clear about resolving it or working around it?

**Comment by rinne** [comment]
@HaleTom Lack of interest from the upstream developers' side.

**Comment by keithw** [comment]
Please see mailing list discussion on this (https://www.mail-archive.com/mosh-devel@mit.edu/msg00980.html).

Excerpt:
> The issue is basically the same since the original pull request in 2013 -- whatever change we make to the Mosh protocol to support ssh-agent forwarding is one we have to live with forever, and the limitations of the Mosh protocol make us not want to commit ourselves to these changes. Mosh does not handle big Instructions well; our fragmentation system is very simple, so adding reliable transport of not-exactly bounded OOB data in the synchronized SSP object makes me nervous.

There is some good news: we think we've figured out a way to do secure ssh-agent forwarding (something that could be enabled all the time, for any server, without the stern warning that accompanies `ssh -A)` and with a prototype that doesn't require changing the core Mosh protocol. Please stay tuned...

**Comment by rinne** [comment]
@keithw 
> we think we've figured out a way to do secure ssh-agent forwarding 

Would you mind sharing?

**Comment by HaleTom** [comment]
I read [the mailing list discussion on this](https://www.mail-archive.com/mosh-devel@mit.edu/msg00980.html) and would appreciate anyone subscribed to that list posting here when the following tool is made available:

> We have developed something internally (at Stanford) that you might
like that also does "secure" ssh-agent forwarding, by allowing the agent to
authenticate and limit (1) the host making the request, (2) the remote host
that host is trying to authenticate to, and (3) the command the host wants
to execute on the remote host. (With normal ssh-agent forwarding, the agent
can't learn any of these things and is basically signing a blank check.)
This works alongside SSH and Mosh. We hope to have a public beta soon and
will look forward to reports from anybody who wants to test it.

**Comment by HaleTom** [comment]
Hmm, is there any work-around for this issue until @keithw 's good news gets implemented (fingers crossed).

**Comment by rinne** [comment]
@HaleTom 
> We have developed something internally (at Stanford) ...

In the dawn of time when there was a standardisation effort for ssh protocol, there was also some new proposals for agent protocol. However, the protocol has remained much the same and there is no real standard for that. The pretty obvious extension that would allow policy based decisions in agent, would be a "forward notification" packet that all _clients_ would add to the forwarded agent connection (towards the agent) and include the information of the endpoint where that particular client is connected. That would mean that agent receiving an agent connection would get 0..n forward notification packets before anything else and could reconstruct the forwarding path before deciding whether to allow signatures and which keys or certificates to show etc.

Basically the user could then limit the agent forwarding to the systems he trusts. And the first level of forward notifications, would only need support in the client the user is using first. Server support is never needed.

Anyways, this is not mosh specific.

**Comment by keithw** [comment]
Hi folks,

Here is our prototype for secure agent forwarding that works with Mosh (or SSH): [Guardian Agent](https://github.com/StanfordSNR/guardian-agent).

Compared with traditional `ssh-agent` forwarding, Guardian Agent provides more-constrained agent forwarding that we think could safely be enabled on any connection. Like Mosh, this tool has to be installed on your local machine and on any host you want to Mosh to. It does not require anything to be installed (or any changes at all) on remote servers (e.g. GitHub) that you want **that** host to be able to SSH to using forwarded credentials. 

This approach also does not require changes to the underlying Mosh protocol to carry ssh-agent request data within the interactive terminal, and we don't have to commit to (and support) a protocol for carrying reliable agent traffic between mosh-clients and mosh-servers. Right now, Guardian Agent works **alongside** any version of Mosh or SSH. Users run `sga-guard` (the agent) on their local machine, in a separate window alongside the interactive session. `sga-guard` prompts the user to approve forwarded `ssh` requests from the intermediary host, either with an X11 popup or in that second terminal window. Based on feedback to this beta/prototype, we may incorporate the techniques behind Guardian Agent more deeply into Mosh.

There is a more detailed writeup [in the README](https://github.com/StanfordSNR/guardian-agent). Please let us know what you think, both about the usability of the tool or about the underlying mechanism. Please file any issues with the tool as GitHub issues in the [Guardian Agent repository](https://github.com/StanfordSNR/guardian-agent). We're looking forward to everybody's feedback.

**Comment by sbz** [comment]
Same for me it doesn't work on version `mosh 1.3.2 [build mosh 1.3.2]`. I tried the following:

```
mosh --ssh="ssh -A" <server>
mosh --ssh="ssh -o ForwardAgent=yes" <server>
```

I can see the `SSH_AUTH_SOCK` variable is exported in the environment, but the agent forwarding doesn't seems to work.

SSH agent forwarding is a very nice feature, I would be happy to see that feature supported by mosh.

**Comment by keithw** [comment]
@sbz Something like that is not going to work -- Mosh only uses SSH to authenticate and remotely invoke the `mosh-server` program. The SSH connection ends and is replaced with Mosh end-to-end. There's no SSH around to forward the agent requests.

If you want ssh-agent forwarding, use Guardian Agent! It works (and is more secure anyway).

**Comment by rinne** [comment]
I am very skeptic about a new agent implementation that requires a script wrapper around ssh is going to see any wide audience adoption. And an agent is already integrated to default sessions in most Linux distros and MacOS.
To that note, I've once more merged the current mosh master branch to my agent forwarding enabled version of mosh at https://github.com/rinne/mosh

**Comment by sbz** [comment]
@keithw Ok thank for your reply and clarification. I will try Guardian Agent!

@rinne I will checkout your branch, looks like the good commit to watch is rinne/mosh@dfa183d

**Comment by tnguyen14** [comment]
@keithw appreciate all the work you and your team put into mosh and Guardian Agent. I think it's great that there's a more secure option for agent forwarding. However, I agree with @rinne that wide adoption is rather unlikely, at least in the short term.

Now that there's a supposedly better solution for agent forwarding, I think mosh should also add support for the builtin agent forwarding, and let the users decide which one they want to use. I think it's detrimental to mosh to not merge #696. Clearly there is a lot of interest and support for that feature. Let the users decide which alternatives they prefer.

**Comment by webbp** [comment]
I would like this feature, too, but a <30 second workaround for connecting to a remote server and then accessing github is to just generate a new ssh key and add it to github, e.g.,
```
ssh-keygen -t rsa -b 4096 -C "your@email.com"
# hit return several times
cat ~/.ssh/id_rsa.pub
# copy it, go here: https://github.com/settings/ssh/new, paste

**Comment by rdnetto** [comment]
> Q: Is Guardian Agent secure?

> A: Guardian Agent is a beta that was first released for beta testing in October 2017. It has not accumulated enough testing and scrutiny to make claims that the implementation is bulletproof.

IMO, this is a pretty strong signal that we should *not* use Guardian Agent for security-conscious environments, which are precisely the ones where we'd prefer agent forwarding over simply storing a key on the remote host.

I appreciate the desire to avoid the issues associated with agent forwarding, but until Guardian Agent is considered stable and secure, I don't think it's a viable alternative.

**Comment by keithw** [comment]
:-) We had a pretty similar disclaimer for the first few years of Mosh, and the results speak for themselves. There's no point in boasting about security before we have a track record. But I think the idea is that if you all try Guardian Agent and the feedback is good, we'll integrate it more tightly into Mosh. That sort of solves the problem of adoption.

**Comment by keithw** [comment]
In general, I know it's frustrating to have a PR sitting around and not get merged, but it's not just a question of the implementation -- we really don't want to complicate the protocol in a way that we then have to support forever going forward (especially by bolting it into the object that gets synchronized by SSP, which is unfortunately a bit of a kludge -- not Timo's fault, it's just that SSP is not very good at this use case).

If you look at where the security holes come from in OpenSSH and OpenSSL, it's largely from added complexity that wasn't _really_ necessary. **Mosh has never had a significant security hole** (yet -- knock on wood) and I don't think that's entirely luck -- it also comes from saying no to things that add complexity to the protocol in a way that makes it harder to reason about.

**Comment by rinne** [comment]
@keithw, just please mark the protobuf message numbers used in agent forwarding patch as reserved in upstream master, so they are not accidentally recycled later. Apart from that, I'm pretty much beyond caring.

**Comment by keithw** [comment]
Sounds pretty reasonable -- you want Mosh to keep tag 8 (your `oob`) reserved in TransportBuffers::Instruction?

**Comment by rinne** [comment]
@keithw "Sounds pretty reasonable -- you want Mosh to keep tag 8 (your oob) reserved in TransportBuffers::Instruction?"

Yes, please. As far as I understand, that should be enough to never break interoperability with agent enabled client vs non-enabled server and vice versa.

**Comment by xhalo32** [comment]
+1 add agent forwarding

**Comment by zviratko** [comment]
This discussion is a bit ridiculous.
It's true that enabling ssh-agent (without confirmation) is a security risk and could be considered bad practice.
But tons of people are doing it, using it and without it, mosh is usable just for their IRC screen hosts (like me).
To install it in an enterprise and use it to manage your servers, you usually _need_ ssh-agent support because sooner or later you'll hit a script that needs it (for git, automation, whatever).

mosh is either a supposed SSH replacement intended for use instead of SSH everywhere possible (in which case it will need to support all the silly stuff as well) or it is just a wrapper that makes roaming easier for few hosts (in which case you might argue improving security is a thing).
I say let the users decide.
I believe mosh security is on-par if not better than SSH with all it's legacy bloat, which should be one more reason to make it possible for people to actually use it as SSH replacement = ssh-agent is a prerequisite of that (and for the record I believe port forwarding as well)

**Comment by keithw** [comment]
Hi folks -- I'm going to lock the conversation here because it doesn't seem to be productive anymore. We are not making the protocol changes proposed in Timo Rinne's pull request, for the reasons we've explained earlier here: we really don't want to complicate the protocol in a way that we then have to support forever going forward (especially by bolting it into the object that gets synchronized by SSP, which is unfortunately a bit of a kludge -- not Timo's fault, it's just that SSP is not very good at this use case).

If you look at where the security holes come from in OpenSSH and OpenSSL, it's largely from added complexity that wasn't really necessary. Mosh has never had a significant security hole (yet -- knock on wood) and I don't think that's entirely luck -- it also comes from saying no to things that add complexity to the protocol in a way that makes it harder to reason about.

If you want ssh-agent forwarding with Mosh, please use Guardian Agent and provide feedback if you have it: https://github.com/StanfordSNR/guardian-agent

It works and I use it every day! If feedback is good, we may consider integrating Guardian Agent more closely into a future version of Mosh. It's true that Guardian Agent is newer and less-tested software, but, so was Mosh when it was younger. The only way it will be more-tested is if more people use it. And, we think it is a lot more secure than "uncontrolled" ssh-agent forwarding.
